helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/SoftConstraint.java

```diff
@@ -27,23 +27,31 @@
  * Evaluate a partition allocation proposal and return a score within the normalized range.
  * A higher score means the proposal is more preferred.
  */
-public interface SoftConstraint {
-  float MIN_SCORE = -1000.0f;
-  float MAX_SCORE = 1000.0f;
+public abstract class SoftConstraint {
+  private Type _type;
 
-  /**
-   * The scoring function returns a score between MINIMAL_SCORE and MAXIMUM_SCORE, which is then weighted by the
-   * individual normalized constraint weights.
-   * Each individual constraint will define the meaning of MINIMAL_SCORE to MAXIMUM_SCORE differently.
-   */
-  float assignmentScore(AssignableNode node, AssignableReplica rep, ClusterContext clusterContext);
+  public SoftConstraint(Type type) {
+    _type = type;
+  }
+
+  enum Type {
+    LEAST_USED_NODE,
+    LEAST_PARTITION_COUNT,
+    LEAST_MOVEMENTS
+  }
 
   /**
-   * Set the importance factor of the soft constraint.
-   * The more important it is, the more contribution it will make to the final evaluation.
-   * @param importance
+   * The scoring function returns a score between MINIMAL_SCORE and MAXIMUM_SCORE, which is then
```

Comment:
Where is the definition of MINIMAL_SCORE and MAXIMUM_SCORE? What will be the exception if this method returns any value out of the scope?
Please note in the comment.

Reply:
Just a comment wouldn't stop us from implementing any constraint at their will. A stronger constraint would be to allow us to implement any ranged soft constraints and the scaler scales the value to a reasonable range plus the weight. 

The MIN MAX scores are just copied from previous soft constraints' interfaces. How about just use Float.MIN_VALUE and Float.MAX_VALUE?

Reply:
That will make the scale function difficult to implement. We need a clear definition.
For now, comments or static vars should be enough. No one can really extend the constraint and deploy except for ourselves.

Reply:
How about adding one more abstract method in SoftConstraint class. Child class either creates their own or inherit the base scale function from the abstract base class.
`ScaleFunction getScaleFunction();` 
Then the `WeightModel` class will basically
```
foreach softconstraint s:
     score = s.getScaleFunction().apply(s.getscore());
```
It's cleaner IMO and allow each child class to customize own its own.

Reply:
It is clean, but not addressing all problems.

If 2 constraints return with completely different scale scores (0-1 and 0-10000 for example), we expect the scale function does the magic and makes these scores comparable. But it might be difficult. So MAYBE, an easier way is just to comment and let the developers of the constraints to do the right thing.

Reply:
Let the developers aware of the scaler function at the interface methods is usually better enforcement than asking for people to reference the comments.

Reply:
Agreed. In that case, can we enforce the input in the scale function? So we have some control on the evaluated score.
Anyway, please try to implement your proposal, we can discuss with the code.